lib/elixir/lib/enum.ex

```diff
@@ -723,28 +723,30 @@ defmodule Enum do
   end
 
   def fetch(enumerable, index) when is_integer(index) and index < 0 do
-    case Enumerable.count(enumerable) do
-      {:error, _module} ->
+    module = Enumerable.impl_for!(enumerable)
+    case module.count(enumerable) do
```

Comment:
First, the documentation of the enumerable protocol talks about returning `{:error, __MODULE__}` only. To me, this means returning anything else than the protocol module is not allowed.

Second, without this, we are doing double dispatch:
* when `Enumerable.count/1` does a dispatch and call returns `{:ok, count}`.
* in `fetch_enumerable/3` we pass `Enumerable`
* `fetch_enumerable/3` does another dispatch.

The only time a double-dispatch does not happen in this function is when count returns `{:error, __MODULE__}`.